packages/flutter/lib/src/services/text_input.dart

```diff
@@ -704,6 +722,387 @@ class RawFloatingCursorPoint {
+    final String newText = textStart + tb + textEnd;
+
+    final bool isEqual = oldText == newText;
+
+    final bool isDeletionGreaterThanOne = (end - start) - (tbEnd - tbStart) > 1;
+    final bool isDeletingByReplacingWithEmpty = tb.isEmpty && tbStart == 0 && tbStart == tbEnd;
+
+    final bool isReplacedByShorter = isDeletionGreaterThanOne && (tbEnd - tbStart < end - start);
+    final bool isReplacedByLonger = tbEnd - tbStart > end - start;
+    final bool isReplacedBySame = tbEnd - tbStart == end - start;
+
+    final bool isInsertingInsideComposingRegion = start + tbEnd > end;
+    final bool isDeletingInsideComposingRegion =
+        !isReplacedByShorter && !isDeletingByReplacingWithEmpty && start + tbEnd < end;
+
+    String newComposingText;
+    String originalComposingText;
+
+    if (isDeletingByReplacingWithEmpty || isDeletingInsideComposingRegion || isReplacedByShorter) {
+      newComposingText = tb.substring(tbStart, tbEnd);
+      originalComposingText = oldText.substring(start, start + tbEnd);
+    } else {
+      newComposingText = tb.substring(tbStart, tbStart + (end - start));
+      originalComposingText = oldText.substring(start, end);
+    }
+
+    final bool isOriginalComposingRegionTextChanged = !(originalComposingText == newComposingText);
+    final bool isReplaced = isOriginalComposingRegionTextChanged ||
+        (isReplacedByLonger || isReplacedByShorter || isReplacedBySame);
+
+    if (isEqual) {
+      return TextEditingDeltaNonTextUpdate(
+        oldText: oldText,
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if ((isDeletingByReplacingWithEmpty || isDeletingInsideComposingRegion) &&
+        !isOriginalComposingRegionTextChanged) {  // Deletion.
+      int actualStart = start;
+
+      if (!isDeletionGreaterThanOne) {
+        actualStart = end - 1;
+      }
+
+      return TextEditingDeltaDeletion(
+        oldText: oldText,
+        deltaText: tb,
+        deltaRange: TextRange(
+          start: actualStart,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if ((start == end || isInsertingInsideComposingRegion) &&
+        !isOriginalComposingRegionTextChanged) {  // Insertion.
+      return TextEditingDeltaInsertion(
+        oldText: oldText,
+        deltaText: tb.substring(end - start, (end - start) + (tb.length - (end - start))),
+        deltaRange: TextRange(
+          start: end,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if (isReplaced) {  // Replacement.
+      return TextEditingDeltaReplacement(
+        oldText: oldText,
+        deltaText: tb,
+        deltaRange: TextRange(
+          start: start,
+          end: end,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+
+    assert(false);
```

Comment:
So this should never be reached, but if it is reached in prod, then use a TextEditingDeltaNonTextUpdate?  Maybe explain this in the message of the assert.

Reply:
Will do